mongo/client_examples_test.go

```diff
@@ -468,3 +469,201 @@ func ExampleConnect_bSONOptions() {
+		opts := options.Client().ApplyURI(uri)
+		opts.SetAuth(
+			options.Credential{
+				Username:                "<client_id>",
+				AuthMechanism:           "MONGODB-OIDC",
+				AuthMechanismProperties: props,
+			},
+		)
+		c, err := mongo.Connect(context.TODO(), opts)
+		if err != nil {
+			panic(err)
+		}
+		defer c.Disconnect(context.TODO())
+		c.Database("test").Collection("test").InsertOne(context.TODO(), bson.D{})
+	}
+
+	// If the application is running on an Azure VM and only one managed identity is associated with the
+	// VM, "username" can be omitted.
+
+	// GCP IMDS
+
+	// For an application running on an GCP VM or otherwise using the `GCP Internal Metadata Service`_,
+	// you can use the built-in support for GCP, where "<audience>" below is the url-encoded "audience"
+	// `configured on your MongoDB deployment`_.
+	{
+		uri := os.Getenv("MONGODB_URI")
+		props := map[string]string{"ENVIRONMENT": "gcp", "TOKEN_RESOURCE": "<audience>"}
+		opts := options.Client().ApplyURI(uri)
+		opts.SetAuth(
+			options.Credential{
+				AuthMechanism:           "MONGODB-OIDC",
+				AuthMechanismProperties: props,
+			},
+		)
+		c, err := mongo.Connect(context.TODO(), opts)
+		if err != nil {
+			panic(err)
+		}
+		defer c.Disconnect(context.TODO())
+		c.Database("test").Collection("test").InsertOne(context.TODO(), bson.D{})
+	}
+
+	// Custom Callbacks
+
+	// For environments that are not directly supported by the driver, you can use
+	// options.OIDCCallback.
+	// Some examples are given below.
+
+	// AWS EKS
+
+	// For an EKS Cluster with a configured `IAM OIDC provider`_, the token can be read from a path given by
+	// the "AWS_WEB_IDENTITY_TOKEN_FILE" environment variable.
+	{
+		eksCallback := func(_ context.Context, _ *options.OIDCArgs) (*options.OIDCCredential, error) {
+			accessToken, err := os.ReadFile(os.Getenv("AWS_WEB_IDENTITY_TOKEN_FILE"))
+			if err != nil {
+				return nil, err
+			}
+			return &options.OIDCCredential{
+				AccessToken: string(accessToken),
+			}, nil
+		}
+		uri := os.Getenv("MONGODB_URI")
+		props := map[string]string{"ENVIRONMENT": "gcp", "TOKEN_RESOURCE": "<audience>"}
+		opts := options.Client().ApplyURI(uri)
+		opts.SetAuth(
+			options.Credential{
+				AuthMechanism:           "MONGODB-OIDC",
+				AuthMechanismProperties: props,
+				OIDCMachineCallback:     eksCallback,
+			},
+		)
+		c, err := mongo.Connect(context.TODO(), opts)
+		if err != nil {
+			panic(err)
+		}
+		defer c.Disconnect(context.TODO())
+		c.Database("test").Collection("test").InsertOne(context.TODO(), bson.D{})
+	}
+
+	// Other Azure Environments
+
+	// For applications running on Azure Functions, App Service Environment (ASE), or
+	// Azure Kubernetes Service (AKS), you can use the `azidentity package`
+	// (https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/azidentity)
+	// to fetch the credentials. In each case, the OIDCCallback function should return
+	// the AccessToken from the azidentity package.
```

Comment:
I wasn't sure how to actually add an example for this because I didn't want to add the azidentity package to the go driver go.mod file.

Reply:
Yeah, I think what you have makes sense.